build-scripts/version.py

```diff
@@ -0,0 +1,149 @@
+"""Write src/version.h and VERSION.txt from git, if available
+
+All commands emulate the Makefile's "version" target.
+Format is: [TAG] [SHA1][-dirty]
+
+This script accepts arguments on the command line with the same format
+as environment variables:
+    VERSION=<version>
+    ARTIFACT=<artifact>
+    TIMESTAMP=<timestamp>
+
+You could run this script like the following.
+- On Unix: ARTIFACT=linux python version.py
+- On Windows: python version.py ARTIFACT=linux
+"""
+
+from pathlib import Path
+import datetime
+import logging
+import os
+import subprocess
+import sys
+
+
+def is_cwd_root():
+    '''Are we on the top-level source directory?
+
+    Bot src/ and data/ must be there.
+    Note: src/ exists in CMake build directories
+    '''
+    return (Path.cwd() / "src").is_dir() and (Path.cwd() / "data").is_dir()
+
+
+def write_version_h(VERSION=''):
+    text = ("//NOLINT(cata-header-guard)\n"
+            f'#define VERSION "{VERSION}"\n')
+    VERSION_H = None
+    src_version_h = Path("src") / "version.h"
+    try:
+        VERSION_H = open(src_version_h, 'r').read()
+    except FileNotFoundError:
+        pass
+    if VERSION_H != text:
+        open(src_version_h, 'w').write(text)
+        return
+    logging.debug("Skip writing src/version.h")
+
+
+def write_VERSION_TXT(GITSHA=None, TIMESTAMP=None, ARTIFACT=None):
+    url = "https://github.com/CleverRaven/Cataclysm-DDA"
+    if GITSHA:
+        url = f"{url}/commit/{GITSHA}"
+    timestamp = TIMESTAMP or datetime.datetime.now().strftime("%Y-%m-%d-%H%M")
+    with open("VERSION.TXT", 'w') as VERSION_TXT:
+        text = str()
+        text += (f"build type: {ARTIFACT or 'Release'}\n"
+                 f"build number: {timestamp}\n"
+                 f"commit sha: {GITSHA or 'Unknown'}\n"
+                 f"commit url: {url}")
+        VERSION_TXT.write(text)
+
+
+def main():
+    logging.basicConfig(level=logging.DEBUG)  # DEBUG
+
+    VERSION = ARTIFACT = TIMESTAMP = GITSHA = None
+
+    # Not using argparse
+    for arg in sys.argv[1:]:
+        arg = arg.split('=', 1)
+        logging.debug(f"{arg}")
+        if len(arg) == 2:
+            if arg[0] == 'ARTIFACT':
+                ARTIFACT = arg[1] or os.environ.get(
+                    'ARTIFACT', None) or 'Release'
+            elif arg[0] == 'TIMESTAMP':
+                TIMESTAMP = arg[1] or os.environ.get('TIMESTAMP', None)
+            elif arg[0] == 'VERSION':
+                VERSION = arg[1] or os.environ.get('VERSION', None)
+            elif arg[0] == 'GITSHA':
+                GITSHA = arg[1] or os.environ.get('GITSHA', None)
+
+    while not is_cwd_root():
+        # Assuming we started somewhere down, climb up to the source directory
+        os.chdir(Path.cwd().parent)
```

Comment:
Resurrecting the previous discussion (it got marked resolved and i missed it):
> We know the location of `version.py` but we don't know from where it will be called

Do we want to call `version.py` from outside the repo?
If yes, then this is incorrect because it will cd you into drive root (and raise exception).
If no, then we already know the repo root because we know where `version.py` is relative to it, and we can cd there directly.

I'm missing something. What failure mode is there with `os.chdir(Path(__file__).parent)`  that `while not is_cwd_root(): os.chdir(Path.cwd().parent)` helps avoid?